sdks/java/io/google-ads/src/main/java/org/apache/beam/sdk/io/googleads/GoogleAdsV14.java

```diff
@@ -0,0 +1,647 @@
+      checkArgumentNotNull(rateLimitPolicyFactory, "rateLimitPolicyFactory cannot be null");
+
+      return toBuilder().setRateLimitPolicyFactory(rateLimitPolicyFactory).build();
+    }
+
+    @Override
+    public PCollection<GoogleAdsRow> expand(PCollection<SearchGoogleAdsStreamRequest> input) {
+      GoogleAdsOptions options = input.getPipeline().getOptions().as(GoogleAdsOptions.class);
+
+      checkArgument(
+          options.getGoogleAdsDeveloperToken() != null || getDeveloperToken() != null,
+          "either --googleAdsDeveloperToken or .withDeveloperToken() is required");
+
+      return input.apply(ParDo.of(new ReadAllFn(this)));
+    }
+
+    /**
+     * A {@link DoFn} that reads reports from Google Ads for each query using the {@code
+     * SearchStream} method.
+     */
+    @VisibleForTesting
+    static class ReadAllFn extends DoFn<SearchGoogleAdsStreamRequest, GoogleAdsRow> {
+      private static final int MAX_RETRIES = 5;
+      private static final FluentBackoff BACKOFF =
+          FluentBackoff.DEFAULT
+              .withExponent(2.0)
+              .withInitialBackoff(Duration.standardSeconds(30))
+              .withMaxRetries(MAX_RETRIES);
+
+      @VisibleForTesting static Sleeper sleeper = Sleeper.DEFAULT;
+
+      private final GoogleAdsV14.ReadAll spec;
+
+      private transient @Nullable GoogleAdsClient googleAdsClient;
+      private transient @Nullable GoogleAdsServiceClient googleAdsServiceClient;
+      private transient @Nullable RateLimitPolicy rateLimitPolicy;
+
+      ReadAllFn(GoogleAdsV14.ReadAll spec) {
+        this.spec = spec;
+      }
+
+      @Setup
+      public void setup(PipelineOptions options) {
+        GoogleAdsOptions adsOptions = options.as(GoogleAdsOptions.class);
+
+        googleAdsClient =
+            spec.getGoogleAdsClientFactory()
+                .newGoogleAdsClient(
+                    adsOptions, spec.getDeveloperToken(), null, spec.getLoginCustomerId());
+        googleAdsServiceClient = googleAdsClient.getVersion14().createGoogleAdsServiceClient();
+        rateLimitPolicy = spec.getRateLimitPolicyFactory().getRateLimitPolicy();
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext c) throws IOException, InterruptedException {
+        GoogleAdsClient googleAdsClient = checkStateNotNull(this.googleAdsClient);
+        GoogleAdsServiceClient googleAdsServiceClient =
+            checkStateNotNull(this.googleAdsServiceClient);
+        RateLimitPolicy rateLimitPolicy = checkStateNotNull(this.rateLimitPolicy);
+
+        BackOff backoff = BACKOFF.backoff();
+        BackOff nextBackoff = backoff;
+        GoogleAdsException lastException = null;
+
+        SearchGoogleAdsStreamRequest request = c.element();
+        String developerToken = googleAdsClient.getDeveloperToken();
+        String customerId = request.getCustomerId();
+
+        do {
+          rateLimitPolicy.onBeforeRequest(developerToken, customerId, request);
```

Comment:
What are the consequences for breaking the rate limit? I ask, because the default rate limiter won't scale with parallel reads.

Reply:
A whole lot of trouble and non-compliance fees, see https://developers.google.com/google-ads/api/docs/rate-sheet.
I've given this a bit more thought and I believe that the policy configuration needs to be owned entirely by the user without giving the illusion of safe defaults, since there simply are none that we can safely recommend due to the complexity of the APIs quotas and limits and their global applicability to developer tokens or customer IDs.

I will rename `DefaultRateLimitPolicy` to `FixedRateLimitPolicy`, mark it public, and leave `rateLimitPolicyFactory` set to null on construction. I'll mark FixedRateLimitPolicy as clearly not intended for advanced users who may exceed their global quota across multiple running applications.

Advanced users should implement their own shared token bucket service (e.g. using https://github.com/mailgun/gubernator) as per the advice provided at https://developers.google.com/google-ads/api/docs/best-practices/rate-limits#throttle.

Moving DatastoreIO's AdaptiveThrottler to a util package would offer a helpful building block for users to apply an additional layer of client side protection against exceeding limits. I'm planning a follow up PR to do that.

Upon reviewing DatastoreIO and FirestoreIO it occurred to me that neither of them apply a pipeline global rate limit across all workers, but divide the desired pipeline global QPS by maxNumWorkers, with their default being 1 QPS local to the DoFn.

Reply:
Done, no policy is set by default but required nonetheless and DefaultRateLimitPolicy has been renamed to SimpleRateLimitPolicy and simply wraps the Guava RateLimiter.
Adequate warnings have been added to the documentation to inform users they should not rely on SimpleRateLimitPolicy for anything but low volume and development use cases.

Reply:
Done.